contributors/design-proposals/auditing.md

```diff
@@ -0,0 +1,290 @@
+
+If auditing is enabled, the http handler will attach an `audit.Event` to the context:
+
+```go
+func WithAuditEvent(parent Context, e *audit.Event) Context
+func AuditEventFrom(ctx Context) (*audit.Event, bool)
+```
+
+Depending on the audit level (see below), different layers of the apiserver (e.g. http handler, storage) will fill the `audit.Event` struct. Certain fields might stay empty or `nil` if given level does not require that field. E.g. in the case when only http headers are supposed to be audit logged, no `OldObject` or `NewObject` is to be retrieved on the storage layer.
+
+Proposed audit levels are:
+
+- `HttpHeader` - reflects the current level of auditing, iow. provides following information about each request: timestamp, source IP, HTTP method, user info (including group as user and as group), namespace, URI and response code.
+- `RequestObject` - additionally provides unstructured object that was part of the request.
+- `StorageObject` - provides the object before and after modification or the difference of the changes (to be decided).
+
+In addition to the amount of details expressed as an audit level, audit filters allows limiting the scope of auditing operations more granularly.
+
+Proposed filters include:
+- user - logs information coming only from specified user(s).
+- namespace - logs information targeting only specified namespace(s).
+- http method - logs information performed only by specified http method(s).
+- resource - logs information targeting specified resource type.
+
+Specified filters will be AND-ed together when processing requests.
+
+When the http request is processed, the request handler will close the audit event by filling in the http response. Then it will pass the event to the configured output backend.
+
+**Note:** for service creation and deletion there is special REST code in the apiserver which takes care of service/node port (de)allocation and removal of endpoints on service deletion. Hence, these operations are not visible on the API layer and cannot be audit logged therefore. **No other resources** (with the exception of componentstatus which is not of interest here) **implement this kind of custom CRUD operations.**
+
+### Events
+
+```go
+package audit
+
+// AuditLevel defines the amount of information logged during auditing
+type AuditLevel string
+
+// Valid audit levels
+const (
+    // AuditHttpHeader provides basic level of auditing, logging data at HTTP level
+    AuditHttpHeader AuditLevel   = "HttpHeader"
+    // AuditRequestobject provides basic level of auditing, and additionally
+    // logs unstructured request object
+    AuditRequestObject AuditLevel = "RequestObject"
+    // AuditStorageobject provides basic and request level, and additionally
+    // logs object before and after saving in storage
+    AuditStorageObject AuditLevel = "StorageObject"
+)
+
+type Event struct {
```

Comment:
What do you think about adding annotations to audit Events? To give a concrete example:

We would like assocatiate a ticket # with every action a production engineer takes when debugging or repairing a customer cluster. We still need to figure out how such annotations would be populated, but it could be implemented as something like an admission controller (if admission controllers get audit events) that looks for audit headers in the request. Another usecase might be a custom plugin extending the audit event with it's own info.